File: tools/vision-review/frame-diff.mjs
#!/usr/bin/env node
// Frame-diff determinista: cuenta píxeles cambiados y su bbox entre frames CONSECUTIVOS de una
// secuencia. Es la referencia directa para separar **movimiento** (cambian las zonas que se
// desplazan) de **glitch** (cambia una zona que debería ser estable). Complementa la detección
// por optical flow (temporal-detect.py) y es lo que prevalece ante una respuesta dudosa del VLM.
//
// Uso: node tools/vision-review/frame-diff.mjs --sequence <dir> [--thresh 40] [--json]
// Salida: imprime una línea por par de frames; con --json, objeto con `pairs`.

import * as fs from 'node:fs';
import * as path from 'node:path';
import { createRequire } from 'node:module';
import { execFileSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';
const require = createRequire(import.meta.url);
const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');

// Punto de entrada unificado: si hay Python + OpenCV, delega en `frame-diff.py` (NumPy/OpenCV, SIMD
// y SSIM). Si no, usa esta implementación Node (pngjs). Así hay una sola verdad para el usuario.
if (!process.argv.includes('--node') && !process.env.FRAME_DIFF_FORCE_NODE) {
  try {
    execFileSync(process.env.PYTHON || 'python', ['-c', 'import cv2, numpy'], { stdio: 'ignore' });
    const py = path.join(ROOT, 'tools/vision-review/frame-diff.py');
    const args = process.argv.slice(2).filter((a) => a !== '--node');
    execFileSync(process.env.PYTHON || 'python', [py, ...args], { stdio: 'inherit' });
    process.exit(0);
  } catch { /* sin Python: sigue con la implementación Node */ }
}

const arg = (n, fb) => { const i = process.argv.indexOf(n); return i >= 0 && i + 1 < process.argv.length ? process.argv[i + 1] : fb; };
const has = (n) => process.argv.includes(n);

const seqDir = arg('--sequence', '');
if (!seqDir) { console.error('Uso: node tools/vision-review/frame-diff.mjs --sequence <dir> [--thresh 40] [--json]'); process.exit(2); }
const thresh = parseInt(arg('--thresh', '40'), 10);

let PNG;
try { ({ PNG } = require('pngjs')); } catch { console.error('[frame-diff] pngjs no disponible.'); process.exit(2); }
const files = fs.readdirSync(seqDir).filter((f) => /^frame_\d{3,}\.png$/.test(f)).sort();
if (files.length < 2) { console.log(`[frame-diff] ${seqDir}: <2 frames (se omite).`); process.exit(3); }
const imgs = files.map((f) => PNG.sync.read(fs.readFileSync(path.join(seqDir, f))));
const W = imgs[0].width, H = imgs[0].height;

const pairs = [];
for (let f = 1; f < imgs.length; f++) {
  const a = imgs[f - 1], b = imgs[f];
  let minx = W, miny = H, maxx = -1, maxy = -1, n = 0;
  for (let y = 0; y < H; y++) for (let x = 0; x < W; x++) {
    const i = (y * W + x) * 4;
    const d = Math.abs(a.data[i] - b.data[i]) + Math.abs(a.data[i + 1] - b.data[i + 1]) + Math.abs(a.data[i + 2] - b.data[i + 2]);
    if (d > thresh) {
      n++;
      if (x < minx) minx = x; if (x > maxx) maxx = x;
      if (y < miny) miny = y; if (y > maxy) maxy = y;
    }
  }
  pairs.push({ from: f - 1, to: f, changed: n, bbox: n ? [minx, miny, maxx, maxy] : null });
}

if (has('--json')) {
  console.log(JSON.stringify({ sequence: seqDir.replace(/\\/g, '/'), size: [W, H], thresh, pairs }, null, 2));
} else {
  console.log(`[frame-diff] ${seqDir.replace(/\\/g, '/')} · ${imgs.length} frames · ${W}×${H} · umbral ${thresh}`);
  for (const p of pairs) {
    const bbox = p.bbox ? `[${p.bbox[0]},${p.bbox[1]}]-[${p.bbox[2]},${p.bbox[3]}]` : '(sin cambio)';
    console.log(`  f${String(p.from).padStart(2, '0')}->f${String(p.to).padStart(2, '0')}: px=${String(p.changed).padStart(6)} ${bbox}`);
  }
}
process.exit(0);

File: tools/vision-review/frame-diff.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Frame-diff y SSIM por bloques, vectorizados (NumPy/OpenCV usan SIMD: SSE/AVX).

Referencia determinista para separar **movimiento** de **glitch** y para medir cambio
**estructural** (contenido) frente a cambio de brillo. Sustituye el bucle por píxel de
`frame-diff.mjs` con operaciones de array (mucho más rápido) y añade:

  - `diff`  : píxeles cambiados (|ΔR|+|ΔG|+|ΔB| > thr) y su bbox, por par de frames.
  - `ssim`  : SSIM medio por par (1.0 = idéntico) y nº de bloques con SSIM bajo (cambio
              estructural real, no solo brillo/desplazamiento).
  - `both`  : las dos métricas.

Uso:
    python tools/vision-review/frame-diff.py --sequence <dir> [--thresh 40]
                                             [--metric diff|ssim|both] [--block 16] [--json]
Salida: por pantalla y `out/vision-review/<demoId>/frame-diff.{json,md}`. Código de salida:
    0 = ok, 2 = uso/entrada, 3 = secuencia insuficiente.

Requiere `opencv-python` y `numpy` (dependencia opcional; ver docs/build/BUILD_AND_RUN.md).
"""
import argparse
import json
import os
import sys
from pathlib import Path

import cv2
import numpy as np


def load_frames(folder, max_frames=None):
    """Carga frames PNG de una carpeta, ordenados por nombre."""
    files = sorted(Path(folder).glob("frame_*.png"))
    frames = []
    for f in (files[:max_frames] if max_frames else files):
        img = cv2.imread(str(f))
        if img is not None:
            frames.append(img)
    return files, frames


def diff_pair(a, b, thr):
    """Píxeles cambiados (|ΔR|+|ΔG|+|ΔB| > thr) y bbox. Vectorizado (SIMD)."""
    # int16 para evitar desbordes al sumar canales.
    d = np.abs(a.astype(np.int16) - b.astype(np.int16)).sum(axis=2)
    mask = d > thr
    n = int(mask.sum())
    if n == 0:
        return {"changed": 0, "bbox": None, "mean": float(d.mean())}
    ys, xs = np.where(mask)
    return {
        "changed": n,
        "bbox": [int(xs.min()), int(ys.min()), int(xs.max()), int(ys.max())],
        "mean": round(float(d.mean()), 3),
    }


def ssim_score(gray_a, gray_b, block):
    """SSIM medio global y nº de bloques con SSIM bajo (cambio estructural).

    SSIM clásico con medias/varianzas gaussianas. `block` px por lado para el recuento
    de bloques estructuralmente distintos (evita que un pequeño cambio domine el global).
    """
    a = gray_a.astype(np.float64)
    b = gray_b.astype(np.float64)
    C1 = (0.01 * 255.0) ** 2
    C2 = (0.03 * 255.0) ** 2
    k = (11, 11)
    mu_a = cv2.GaussianBlur(a, k, 1.5)
    mu_b = cv2.GaussianBlur(b, k, 1.5)
    mu_a2, mu_b2, mu_ab = mu_a * mu_a, mu_b * mu_b, mu_a * mu_b
    sig_a2 = cv2.GaussianBlur(a * a, k, 1.5) - mu_a2
    sig_b2 = cv2.GaussianBlur(b * b, k, 1.5) - mu_b2
    sig_ab = cv2.GaussianBlur(a * b, k, 1.5) - mu_ab
    ssim_map = ((2 * mu_ab + C1) * (2 * sig_ab + C2)) / ((mu_a2 + mu_b2 + C1) * (sig_a2 + sig_b2 + C2))
    mean = float(ssim_map.mean())
    # Bloques con SSIM bajo (cambio estructural localizado).
    h, w = ssim_map.shape
    low = 0
    for y in range(0, max(1, h - block + 1), block):
        for x in range(0, max(1, w - block + 1), block):
            if float(ssim_map[y:y + block, x:x + block].mean()) < 0.85:
                low += 1
    return {"ssim": round(mean, 5), "blocks_low": low}


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--sequence", required=True)
    ap.add_argument("--out", default=None)
    ap.add_argument("--thresh", type=int, default=40)
    ap.add_argument("--metric", choices=["diff", "ssim", "both"], default="both")
    ap.add_argument("--block", type=int, default=16)
    ap.add_argument("--json", action="store_true")
    args = ap.parse_args()

    files, frames = load_frames(args.sequence)
    if len(frames) < 2:
        print(f"[frame-diff] {args.sequence}: <2 frames (se omite).", file=sys.stderr)
        return 3
    h, w = frames[0].shape[:2]

    pairs = []
    for f in range(1, len(frames)):
        entry = {"from": f - 1, "to": f}
        if args.metric in ("diff", "both"):
            entry.update(diff_pair(frames[f - 1], frames[f], args.thresh))
        if args.metric in ("ssim", "both"):
            ga = cv2.cvtColor(frames[f - 1], cv2.COLOR_BGR2GRAY)
            gb = cv2.cvtColor(frames[f], cv2.COLOR_BGR2GRAY)
            entry.update(ssim_score(ga, gb, args.block))
        pairs.append(entry)

    seq_path = Path(args.sequence).resolve()
    demo_id = seq_path.parts[-3] if len(seq_path.parts) >= 3 else seq_path.parent.name
    out_dir = args.out or os.path.join("out", "vision-review", demo_id)
    os.makedirs(out_dir, exist_ok=True)

    report = {
        "sequence": args.sequence.replace("\\", "/"),
        "demo": demo_id,
        "frames": len(frames),
        "size": [w, h],
        "metric": args.metric,
        "thresh": args.thresh,
        "block": args.block,
        "pairs": pairs,
    }
    with open(os.path.join(out_dir, "frame-diff.json"), "w", encoding="utf-8") as fp:
        json.dump(report, fp, indent=2, ensure_ascii=False)

    cols = ["par"]
    if args.metric in ("diff", "both"):
        cols += ["px cambiados", "bbox", "Δ medio"]
    if args.metric in ("ssim", "both"):
        cols += ["SSIM", "bloques bajos"]
    lines = [f"| {' | '.join(cols)} |", "|" + "---|" * len(cols)]
    for p in pairs:
        row = [f"f{p['from']}→f{p['to']}"]
        if args.metric in ("diff", "both"):
            bb = p["bbox"]
            row += [str(p["changed"]), f"{bb[0]},{bb[1]}–{bb[2]},{bb[3]}" if bb else "(sin cambio)", str(p["mean"])]
        if args.metric in ("ssim", "both"):
            row += [str(p["ssim"]), str(p["blocks_low"])]
        lines.append("| " + " | ".join(row) + " |")

    md = "\n".join([
        f"# Frame-diff / SSIM — {demo_id}",
        "",
        f"Secuencia: `{report['sequence']}` · {len(frames)} frames · {w}×{h} px · "
        f"métrica `{args.metric}` · umbral diff {args.thresh} · bloque {args.block}.",
        "",
        *lines,
        "",
        "> `diff` mide píxeles cambiados (movimiento o glitch); `SSIM` mide cambio **estructural** "
        "(1.0 = idéntico). Un cambio localizado con SSIM alto y diff alto suele ser movimiento.",
    ]) + "\n"
    with open(os.path.join(out_dir, "frame-diff.md"), "w", encoding="utf-8") as fp:
        fp.write(md)

    if args.json:
        print(json.dumps(report, ensure_ascii=False))
    else:
        print(f"[frame-diff] {demo_id}: {len(pairs)} pares; informe: {out_dir}/frame-diff.md")
    return 0


if __name__ == "__main__":
    sys.exit(main())
